### PR TITLE
Add document for Provision volumes from cross namespace data sources

### PR DESCRIPTION
**Pull request by ttakahashi21**
Add document for Provision volumes from cross namespace data sources
 
■KEP
Please see [CrossNamespaceVolumeDataSource](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots)

■API Change (Merge)
https://github.com/kubernetes/kubernetes/pull/113186

■Controller (Under Review)
https://github.com/kubernetes-csi/external-provisioner/pull/805

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -689,6 +690,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
   This feature gate guards *a group* of CPUManager options whose quality level is beta.
   This feature gate will never graduate to stable.
 - `CPUManagerPolicyOptions`: Allow fine-tuning of CPUManager policies.
+- `CrossNamespaceVolumeDataSource`: Enable usage of Provision of PVCs from VolumeSnasphot in other namespaces.
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:693]:
```suggestion
- `CrossNamespaceVolumeDataSource`: Enable provisioning a {{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}}
  (PVC) that is based on VolumeSnapshot in a different namespace from the PVC.
```

**Review by sftim — COMMENTED** [review]
Thanks. Please also update https://kubernetes.io/docs/concepts/storage/volume-snapshots/#provisioning-volumes-from-snapshots to document the new optional / alpha feature.

**Comment by sftim** [comment]
BTW:
if this PR is still draft, that's OK - please mark the PR as draft by, eg, editing its title. If it's not draft, please write a PR description.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (ttakahashi21) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: Fixed.

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 2]
```diff
@@ -1010,6 +1010,62 @@ is registered to handle that kind of data source. When a suitable populator is i
 responsibility of that populator controller to report Events that relate to volume creation and issues during
 the process.
 
+## Provision of PersistentVolumeClaims from VolumeSnasphot in other namespaces
+
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
```
Comment (msau42) [inline content/en/docs/concepts/storage/persistent-volumes.md:1015]: I think it would be good to incorporate this section into the "Data source references" above.

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 3]
```diff
@@ -1010,6 +1010,62 @@ is registered to handle that kind of data source. When a suitable populator is i
...
+     name: foo-pvc
+     namespace: ns1
+   spec:
+     storageClassName: example
+     accessModes:
+     - ReadWriteOnce
+     resources:
+       requests:
+         storage: 1Gi
+     dataSourceRef2:
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md:1048]: Fixed.

File: content/en/docs/concepts/storage/persistent-volumes.md (hunk 2 above)
Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md]: I see. Fixed https://github.com/kubernetes/website/pull/37802/commits/ae77a57a131507e2958d4ac11fe290c70392f66b

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 4]
```diff
@@ -970,6 +982,12 @@ users should be aware of:
 * The `dataSourceRef` field may contain different types of objects, while the `dataSource` field
   only allows PVCs and VolumeSnapshots.
 
+In case of enabling the `CrossNamespaceVolumeDataSource` feature gate, there are three important differences between the `dataSource` field and the `dataSourceRef` field that users should be aware of: .
+
+* While the `dataSource` field only allows two specific types of objects, the `dataSourceRef` field allows any non-core object, as well as PersistentVolumeClaim objects.
+* While the `dataSource` field ignores disallowed values (dropping them), the `dataSourceRef` field preserves all values, and generates an error if a disallowed value is specified.
+* While the `dataSource`field only allows local objects, the `dataSourceRef` field allows objects in any namespaces.
```
Comment (msau42) [inline content/en/docs/concepts/storage/persistent-volumes.md:989]: I think the first 2 are repeats of L979-L983.

Probably can reword this to say something like

```
When the CrossNamespaceVolumeDataSource feature is enabled, there are additional differences:
 
* The `dataSource`field only allows local objects, while the `dataSourceRef` field allows objects in any namespaces.  
* `dataSource` and `dataSourceRef` are not synced.
```

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 5]
```diff
@@ -953,6 +953,18 @@ or to a VolumeSnapshot, the `dataSourceRef` field can contain a reference to any
 same namespace, except for core objects other than PVCs. For clusters that have the feature
 gate enabled, use of the `dataSourceRef` is preferred over `dataSource`.
 
+## Cross Namespace Volume Data Source and data sources
```
Comment (msau42) [inline content/en/docs/concepts/storage/persistent-volumes.md:956]: Reword to "Cross namespace data sources"

File: content/en/docs/concepts/storage/persistent-volumes.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (msau42) [reply content/en/docs/concepts/storage/persistent-volumes.md:1048]: I don't see the fix. Did you reupload?

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 6]
```diff
@@ -1010,6 +1028,31 @@ is registered to handle that kind of data source. When a suitable populator is i
 responsibility of that populator controller to report Events that relate to volume creation and issues during
 the process.
 
+### Using Cross Namespace Volume Data Source
+
```
Comment (msau42) [inline content/en/docs/concepts/storage/persistent-volumes.md:1032]: Can you add the feature gate tag here, and also add an example of the ReferenceGrant?

Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md:1032]: Add feature gate tag and example.

File: content/en/docs/concepts/storage/persistent-volumes.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md]: Sorry, I mistake. I fixed now.

File: content/en/docs/concepts/storage/persistent-volumes.md (hunk 4 above)
Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md]: Thank you for suggestion. Fixed https://github.com/kubernetes/website/pull/37802/commits/7cfe0a3f71a8d53f981052ab2c5864c771ce3230

**Comment by msau42** [comment]
tech review lgtm, just some minor nits. Looks like the PR needs to be rebased as well.

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 7]
```diff
@@ -969,6 +981,11 @@ users should be aware of:
 * The `dataSourceRef` field may contain different types of objects, while the `dataSource` field
   only allows PVCs and VolumeSnapshots.
 
+When the `CrossNamespaceVolumeDataSource` feature is enabled, there are additional differences:
+
+* The `dataSource` field only allows local objects, while the `dataSourceRef` field allows objects in any namespaces.  
+* When namespace is specified, dataSource and dataSourceRef are not synced.
```
Comment (msau42) [inline content/en/docs/concepts/storage/persistent-volumes.md:987]: Add the backticks for `dataSource` and `dataSourceRef`

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (ttakahashi21) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: @sftim 
https://github.com/kubernetes/website/pull/37802/files#diff-f937c4715e7d2fa6db515660503a56c87c317122dc2860873c7022e20e452141R414-R415
I checked the contents of the description again and fixed.

**Comment by leonardpahlke** [comment]
cc @krol3

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 8]
```diff
@@ -952,6 +952,18 @@ or to a VolumeSnapshot, the `dataSourceRef` field can contain a reference to any
 same namespace, except for core objects other than PVCs. For clusters that have the feature
 gate enabled, use of the `dataSourceRef` is preferred over `dataSource`.
 
+## Cross namespace data sources
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
+
+Kubernetes supports cross namespace volume data source.
+To use cross namespace volume datasource, you must enable the `AnyVolumeDataSource` and `CrossNamespaceVolumeDataSource`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) for
```
Comment (sftim) [inline content/en/docs/concepts/storage/persistent-volumes.md:960]:
```suggestion
To use cross namespace volume data sources, you must enable the `AnyVolumeDataSource` and `CrossNamespaceVolumeDataSource`
[feature gates](/docs/reference/command-line-tools-reference/feature-gates/) for
```

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 9]
```diff
@@ -1009,6 +1026,50 @@ is registered to handle that kind of data source. When a suitable populator is i
 responsibility of that populator controller to report Events that relate to volume creation and issues during
 the process.
 
+### Using Cross Namespace Volume Data Source
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
+
+Create a gateway.networking.k8s.io/ReferenceGrant to allow the namespace owner to accept the reference.
+Users create a populated volume by referring cross namespace volume data source using the `dataSourceRef` field:
```
Comment (sftim) [inline content/en/docs/concepts/storage/persistent-volumes.md:1033]:
```suggestion
You define a populated volume by specifying a cross namespace volume data source using the `dataSourceRef` field. You must already have a valid ReferenceGrant in the source namespace:
```

**Review by sftim — COMMENTED** [review]
For beta, we must make sure mention ReferenceGrants in the authorization documentation - see https://kubernetes.io/docs/reference/access-authn-authz/

A mention in https://kubernetes.io/docs/concepts/security/ would be nice-to-have for beta, too.

It'd be even better to make those changes early, for alpha, but it's not mandatory.

**Comment by krol3** [comment]
Hi @ttakahashi21, This PR needs a doc review by **Mon Nov 28th** to get this into the release. Please reach out to required SIGs to get their review. Thank you!
cc: @mickeyboxell

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 10]
```diff
@@ -1009,6 +1026,50 @@ is registered to handle that kind of data source. When a suitable populator is i
...
+
+   ```yaml
+   apiVersion: gateway.networking.k8s.io/v1beta1
+   kind: ReferenceGrant
+   metadata:
+     name: allow-ns1-pvc
+     namespace: default
+   spec:
+     from:
+     - group: ""
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply content/en/docs/concepts/storage/persistent-volumes.md:1043]: Because PersistentVolumeClaim API is core API

**Comment by ttakahashi21** [comment]
@sftim I fixed those that needed fixing.

File: content/en/docs/concepts/storage/persistent-volumes.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/storage/persistent-volumes.md:1050]: (don't tell _me_ - tell the readers!)

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 11]
```diff
@@ -952,6 +952,25 @@ or to a VolumeSnapshot, the `dataSourceRef` field can contain a reference to any
...
+## Cross namespace data sources
+{{< feature-state for_k8s_version="v1.26" state="alpha" >}}
+
+Kubernetes supports cross namespace volume data sources.
+To use cross namespace volume data sources, you must enable the `AnyVolumeDataSource` and `CrossNamespaceVolumeDataSource`
+[feature gates](/docs/reference/command-line-tools-reference/feature-gates/) for
+the kube-apiserver, kube-controller-manager.
+Also, you must enable the `CrossNamespaceVolumeDataSource` feature gate for the csi-provisioner.
+
+Enabling the `CrossNamespaceVolumeDataSource` feature gate allow you to specify a namespace in the dataSourceRef field.
```
Comment (sftim) [inline content/en/docs/concepts/storage/persistent-volumes.md:964]: (nit)
```suggestion
Enabling the `CrossNamespaceVolumeDataSource` feature gate allow you to specify a
namespace in the `dataSourceRef` field of a PersistentVolumeClaim.
```

**Review by sftim — COMMENTED** [review]
Thanks

/approve
